From 1056f579c1612ac873fe0eedd61b0d52bff9ea20 Mon Sep 17 00:00:00 2001
From: buildbot <build@bot.com>
Date: Mon, 23 Nov 2015 22:08:34 -0600
Subject: [PATCH 2/3] Revert "ifdown: really set down all active interfaces"

This doesn't exist in the Android libc.

This reverts commit c75354b291d1c8cc306d02a63372d487299188b3.
---
 kexec/ifdown.c | 42 +++++++++++++++++++++---------------------
 1 file changed, 21 insertions(+), 21 deletions(-)

diff --git a/kexec/ifdown.c b/kexec/ifdown.c
index 46b7bef..9824153 100644
--- a/kexec/ifdown.c
+++ b/kexec/ifdown.c
@@ -19,6 +19,8 @@ char *v_ifdown = "@(#)ifdown.c  1.11  02-Jun-1998  miquels@cistron.nl";
 #include <net/if.h>
 #include <netinet/in.h>
 
+#define MAX_IFS	64
+
 /*
  *	First, we find all shaper devices and down them. Then we
  *	down all real interfaces. This is because the comment in the
@@ -27,45 +29,43 @@ char *v_ifdown = "@(#)ifdown.c  1.11  02-Jun-1998  miquels@cistron.nl";
  */
 int ifdown(void)
 {
-	struct if_nameindex *ifa, *ifp;
-	struct ifreq ifr;
-	int fd, shaper;
+	struct ifreq ifr[MAX_IFS];
+	struct ifconf ifc;
+	int i, fd;
+	int numif;
+	int shaper;
 
 	if ((fd = socket(AF_INET, SOCK_DGRAM, 0)) < 0) {
 		fprintf(stderr, "ifdown: ");
 		perror("socket");
 		return -1;
 	}
+	ifc.ifc_len = sizeof(ifr);
+	ifc.ifc_req = ifr;
 
-	if ((ifa = if_nameindex()) == NULL) {
+	if (ioctl(fd, SIOCGIFCONF, &ifc) < 0) {
 		fprintf(stderr, "ifdown: ");
-		perror("if_nameindex");
+		perror("SIOCGIFCONF");
+		close(fd);
 		return -1;
 	}
+	numif = ifc.ifc_len / sizeof(struct ifreq);
 
 	for (shaper = 1; shaper >= 0; shaper--) {
-		for (ifp = ifa; ifp->if_index; ifp++) {
+		for (i = 0; i < numif; i++) {
 
-			if ((strncmp(ifp->if_name, "shaper", 6) == 0)
+			if ((strncmp(ifr[i].ifr_name, "shaper", 6) == 0)
 			    != shaper) continue;
-			if (strcmp(ifp->if_name, "lo") == 0)
+
+			if (strcmp(ifr[i].ifr_name, "lo") == 0)
 				continue;
-			if (strchr(ifp->if_name, ':') != NULL)
+			if (strchr(ifr[i].ifr_name, ':') != NULL)
 				continue;
-
-			strncpy(ifr.ifr_name, ifp->if_name, IFNAMSIZ);
-			if (ioctl(fd, SIOCGIFFLAGS, &ifr) < 0) {
+			ifr[i].ifr_flags &= ~(IFF_UP);
+			if (ioctl(fd, SIOCSIFFLAGS, &ifr[i]) < 0) {
 				fprintf(stderr, "ifdown: shutdown ");
-				perror(ifp->if_name);
-				return -1;
+				perror(ifr[i].ifr_name);
 			}
-			ifr.ifr_flags &= ~(IFF_UP);
-			if (ioctl(fd, SIOCSIFFLAGS, &ifr) < 0) {
-				fprintf(stderr, "ifdown: shutdown ");
-				perror(ifp->if_name);
-				return -1;
-			}
-
 		}
 	}
 	close(fd);
-- 
2.5.0

